### PR TITLE
read swath window file with no header try #2

### PR DESCRIPTION
**Pull request by jcharkow**
Reopen https://github.com/OpenMS/OpenMS/pull/6070 with alternative branch so that code formatting does not fail

In response to https://github.com/OpenMS/OpenMS/issues/6055 allow
openSwath to be able to read window files with no header.

# Description

Please include a summary of the change and which issue is fixed.

# Checklist:
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes. (Tick if no updates were necessary.)

# How can I get additional information on failed tests during CI:
If your PR is failing you can check out 
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. If you click in the column that lists the failed tests you will get detailed error messages.
- Or click on the action: e.g., for clang-format linting

# Note:
- Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

# Advanced commands (admins / reviewer only):
- `/reformat` (experimental) applies the clang-format style changes as additional commit
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds

### CODE REVIEW

**Comment by timosachsenberg** [comment]
hmm there still seems to be wrong with the reformat script. not sure why it e.g. reformated SiriusAdapterAlgorithm at all

**Comment by timosachsenberg** [comment]
sorry can you revert the last commit by the linter?

**Comment by timosachsenberg** [comment]
I will revert

**Comment by jpfeuffer** [comment]
Probably a GitHub bug since it tagged your comment for a completely different commit. Please raise an issue there or with their customer support

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp [hunk 1]
```diff
@@ -115,10 +115,19 @@ namespace OpenMS
...
     std::ifstream data(filename.c_str());
     std::string line;
-    std::getline(data, line); //skip header
-    std::cout << "Read Swath window header: '" << line << "'\n";
+
+    // check for presence of header, a header is present if the first line contains alphabetic characters
     double lower, upper;
-    while (std::getline(data, line))
+    std::getline(data, line); // possible header (or first line)
+    if (std::any_of(line.begin(), line.end(), [](char const &c) { return std::isalpha(c); })){ // if any of the characters are alphabetical then there is a header
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp:122]:
```suggestion
    if (std::any_of(line.begin(), line.end(), [](char const &c) { return std::isalpha(c); }))
    { // if any of the characters are alphabetical then there is a header
```

**Comment by jcharkow** [comment]
Is this ready to be merged?

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp [hunk 2]
```diff
@@ -115,10 +115,21 @@ namespace OpenMS
...
-    std::getline(data, line); //skip header
-    std::cout << "Read Swath window header: '" << line << "'\n";
+
+    // check for presence of header, a header is present if the first line contains alphabetic characters
     double lower, upper;
-    while (std::getline(data, line))
+    std::getline(data, line); // possible header (or first line)
+    if (std::any_of(line.begin(), line.end(), [](char const &c) { return std::isalpha(c); }))
+    { // if any of the characters are alphabetical then there is a header
+      std::cout << "Read Swath window header: '" << line << "'\n";
```
Comment (jpfeuffer) [inline src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp:124]:
```suggestion
      OPENMS_LOG_INFO << "Read Swath window header: '" << line << "'\n";
```

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp [hunk 3]
```diff
@@ -115,10 +115,21 @@ namespace OpenMS
...
+      std::getline(data, line); // first line of swath window
+     }
+    else 
+    {
+      std::cout << "Swath Header not found'\n";
```
Comment (jpfeuffer) [inline src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp:129]:
```suggestion
      OPENMS_LOG_INFO << "Swath Header not found'\n";
```

**Review by jpfeuffer — COMMENTED** [review]
Check the logging. Unless there are problems with multithreading, please use OPENMS_LOG_...

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp [hunk 4]
```diff
@@ -115,10 +115,21 @@ namespace OpenMS
...
     std::ifstream data(filename.c_str());
     std::string line;
-    std::getline(data, line); //skip header
-    std::cout << "Read Swath window header: '" << line << "'\n";
+
+    // check for presence of header, a header is present if the first line contains alphabetic characters
     double lower, upper;
-    while (std::getline(data, line))
+    std::getline(data, line); // possible header (or first line)
+    if (std::any_of(line.begin(), line.end(), [](char const &c) { return std::isalpha(c); }))
```
Comment (cbielow) [inline src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp:122]: what if I write
```
1e10, 2e10
```
into that file ?! :)

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp [hunk 5]
```diff
@@ -37,6 +37,9 @@
 #include <OpenMS/CONCEPT/Types.h>
 #include <OpenMS/CONCEPT/Exception.h>
 #include <OpenMS/DATASTRUCTURES/String.h>
+#include <OpenMS/SYSTEM/SysInfo.h>
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp:40]: minor: we sort headers by name

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp [hunk 6]
```diff
@@ -114,21 +117,24 @@ namespace OpenMS
...
-    if (std::any_of(line.begin(), line.end(), [](char const &c) { return std::isalpha(c); }))
-    { // if any of the characters are alphabetical then there is a header
-      std::cout << "Read Swath window header: '" << line << "'\n";
-      std::getline(data, line); // first line of swath window
-     }
-    else 
+
+    // Check for presence of header
+    std::getline(data, line);
+    try { // If string can be successfully converted to double (excluding initial spaces) then the first line is not a header
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp:126]: bracket on new line... check out https://github.com/OpenMS/OpenMS/wiki/Coding-conventions

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp [hunk 7]
```diff
@@ -114,21 +117,24 @@ namespace OpenMS
...
+      StringUtils::split(line.trim(), ' ', headerSubstrings);
```
Comment (timosachsenberg) [inline src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp:127]: in OpenMS we mainly use tab separated files not space separated.
Would it make sense to make this consistent here?
or allow both?

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp (hunk 6 above)
Comment (timosachsenberg) [reply src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp]: I added this in a commit

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp (hunk 5 above)
Comment (jpfeuffer) [reply src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp]: By filename or filepath?

Comment (timosachsenberg) [reply src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp]: Path

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp [hunk 8]
```diff
@@ -114,11 +118,26 @@ namespace OpenMS
...
     double lower, upper;
-    while (std::getline(data, line))
+
+    // Check for presence of header
+    std::getline(data, line);
+    try 
+    { // If string can be successfully converted to double (excluding initial spaces) then the first line is not a header
+      StringUtils::split(line.trim(), ' ', headerSubstrings);
+      StringUtils::toDouble(headerSubstrings[0]);
+      std::cout << "Swath Header not found \n";
```
Comment (jpfeuffer) [inline src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp:131]: What happened to OpenMS log? :D

File: src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp (hunk 7 above)
Comment (jcharkow) [reply src/openms/source/ANALYSIS/OPENSWATH/SwathWindowLoader.cpp]: The test file seems to use space separation (as oppose to tabs) so I think we should keep the space separation. I tried to address this issue here https://github.com/OpenMS/OpenMS/pull/6077/commits/10064a21c36cfe3b5e1de57c68b84b121612948a

**Review by timosachsenberg — APPROVED** [review]

**Comment by timosachsenberg** [comment]
Thanks. Looks good from my side.

**Comment by timosachsenberg** [comment]
FYI: https://cdash.openms.de/viewBuildError.php?buildid=36147 I can fix it

**Pull request merged**